### ISSUE TITLE
Automatic runtime install of tools

### ISSUE THREAD

**Issue by rarkins** [containerbase/buildpack#101]
Depends on #16 

An even more advanced approach here would be that we create our own tools directory in `$PATH`, e.g. containing `node`, `poetry`, `gradle`, etc. But each is in fact a wrapper/stub and its purpose is to dynamically install the required tool + language and then pass the same parameters to the tool itself after it's installed.

e.g. the application like Renovate will simply call `poetry` via child process, which then kicks off:
- Our own custom shell script
- Installs `python` if missing (According to `PYTHON_VERSION`)
- Installs `poetry`
- Calls newly installed `poetry` and passes all args to it

**Comment by viceice** [comment]
@rarkins Do we still like to do this?

When such a wrapper is called and a version is passed via env, use that version.
If version is not set use latest version or fail?

**Comment by rarkins** [comment]
Yes, we need this for other products at Mend. Except we also want it to do it's own auto detection too, for the tools which don't have the smarts like Renovate

**Comment by viceice** [comment]
ok, detecting tool version from current dir (until repo root) will be tricky. so i'm sure it's nothing we should try to do in bash scripts.

**Comment by rarkins** [comment]
My plan is to do it first using TS so that we can reuse renovate's datasource lookup logic. It can be compiled to a binary using pkg. After that porting to Go is a nice to have. I don't think this forces the current code to be rewritten in the same time frame and can remain as bash

**Comment by viceice** [comment]
sounds good.

https://github.com/containerbase/base/issues/771#issuecomment-1473279221

**Comment by viceice** [comment]
btw: `PIPENV_VERSION` can't be used[^1]

[^1]: https://github.com/pypa/pipenv/issues/3633#issuecomment-478250721

**Comment by rarkins** [comment]
Can it be unset before we call to pipenv?

**Comment by viceice** [comment]
maybe, not working on it now.
